server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1597,105 +1624,30 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer, Entitlement enti
         Map<String, Integer> poolQuantityMap, CallerType caller)
         throws EntitlementRefusedException {
 
-        // Because there are several paths to this one place where entitlements
-        // are granted, we cannot be positive the caller obtained a lock on the
-        // pool when it was read. As such we're going to reload it with a lock
-        // before starting this process.
-        log.debug("Locking pools: {}", poolQuantityMap.keySet());
-        Collection<Pool> pools = poolCurator.lockAndLoadByIds(poolQuantityMap.keySet());
-
-        if (log.isDebugEnabled()) {
-            for (Pool pool : pools) {
-                log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
-            }
-        }
-
-        log.debug("Done locking pools");
+        BindChain chain = new BindChain();
+        BindContext context = bindContextFactory.create(consumer, poolQuantityMap);
+        chain.addOperation(rulesCheckOpFactory.create(caller));
+        chain.addOperation(handleEntitlementsOp);
+        chain.addOperation(postBindBonusPoolsOp);
+        chain.addOperation(handleCertificatesOp);
+        chain.addOperation(complianceOp);
 
```

Comment:
I don't really like the fact that all the operations bleed into the pool manager. The chain is really the only thing that needs to know about the operations. They shouldn't be getting injected into the pool manager IMO.

Perhaps it would be better to inject the chain in the pool manager and do operation injections inside of the chain itself to isolate the operations to the chain.

Then we can simply:
```java
BindContext context = bindContextFactory.create(consumer, poolQuantityMap);
chain.preProcess(context);
...
```